src/lib.rs

```diff
@@ -73,115 +100,253 @@ fn encode_inner(shards: &mut [&mut [u8]], data_shards: usize, shard_size: usize)
-            .collect::<Vec<_>>();
-        let mut xor_out_chunk = xor_out
-            .iter_mut()
-            .map(|shard| &mut shard[offset..end])
-            .collect::<Vec<_>>();
-
-        // copy the input to the work table
-        for (shard, work) in shards_chunk[data_shards..]
-            .iter()
-            .zip(work_chunk.iter_mut())
-        {
-            work.copy_from_slice(shard);
+    // copy the input to the work table
+    for (shard, work) in shards[data_shards..].iter().zip(work.iter_mut()) {
+        work.copy_from_slice(shard);
+    }
+
+    ifft_dit_encoder(
+        &shards[..data_shards],
+        mtrunc,
+        &mut work,
+        None, // No xor output
+        m,
+        skew_lut,
+    );
+
+    let last_count = data_shards % m;
+
+    // goto skip_body
+    if m < data_shards {
+        let (xor_out, work) = work.split_at_mut(m);
+        let mut n = m;
+
+        // for sets of m data pieces
+        while n <= data_shards - m {
+            // work <- work xor IFFT(data + i, m, m + i)
+            ifft_dit_encoder(&shards[n..], m, work, Some(xor_out), m, &skew_lut[n..]);
+            n += m;
         }
 
-        ifft_dit_encoder(
-            &shards_chunk[..data_shards],
-            mtrunc,
-            &mut work_chunk,
-            None, // No xor output
-            m,
-            skew_lut,
-        );
+        // Handle final partial set of m pieces:
+        if last_count != 0 {
+            ifft_dit_encoder(
+                &shards[n..],
+                last_count,
+                work,
+                Some(xor_out),
+                m,
+                &skew_lut[n..],
+            );
+        }
+    }
+
+    // work <- FFT(work, m, 0)
+    fft_dit(&mut work, parity_shards, m, &*lut::FFT_SKEW);
+
+    for (shard, work) in shards[data_shards..].iter_mut().zip(work.iter()) {
+        shard.copy_from_slice(work);
+    }
+}
+
+/// Leopard algorithm is imperfect and can result in a buffer overflow on the 'lut::FFT_SKEW'.
+/// Encoding happens in passes where each pass encodes parity for data shards in chunks
+/// of the smallest power of 2 bigger (or equal) than parity shards amount.
+/// If the last chunk is not a full pass, we can hit the overflow for some pairs
+/// of (data_shards, parity_shards).
+/// This function detects such inputs.
+fn is_encode_buf_overflow(data_shards: usize, parity_shards: usize) -> bool {
+    debug_assert!(data_shards >= parity_shards);
+    debug_assert!(data_shards + parity_shards <= ORDER);
+
+    let m = ceil_pow2(parity_shards);
+    let last_count = data_shards % m;
+
+    // we can finish encoding with only full passes
+    if m >= data_shards || last_count == 0 {
+        return false;
+    }
+
+    let full_passes = data_shards / m;
+    // if this is 'true', we would overflow the fft skew table which has the size of `MODULUS`
+    (full_passes + 1) * m + 1 > MODULUS as usize
+}
+
+/// Reconstructs the original shards from the provided slice.
```

Comment:
```suggestion
/// Reconstructs original shards from the provided slice.
```
I'm like 80% certain 🤔